### ISSUE TITLE
Feature request: Please drop the direct use of int, short, etc in favour of int_least32_t, int_least16_t

### ISSUE THREAD

**Issue by awsdert** [floooh/sokol#1083]
I'm designing an ABI solution to programming across different platforms and this library looks like it would do nicely as the backend for game related APIs in it. The only thing that complicates that is the inconsistent ABI sokol currently uses by tying itself to the data model instead of using fixed size integers and the like.

I could probably get away with using only CPU register sized integers for the public API part but it would be nicer to use the same sized integers of a future version of sokol that drops the data model reliance in favour of fixed size integers.

For instance `int sapp_gl_get_major_version(...)` could instead become `int_least16_t sapp_gl_get_major_version(...)`. Even if sokol were to somehow reach 30000 odd major versions and need to change to int_least32_t, by that point my library would've also changed ABI and version (because the ABI is tied to the version and accompanying launcher).

**Comment by floooh** [comment]
Are you planning to support platforms where `int` isn't 32-bit, if yes, which?

**Comment by awsdert** [comment]
long can be 32 or 64 bit,
int can be 16, 32 or 64 bit
short can be 16 or 64 bit
void* can be 16, 32 or 64 bit

These details are all part of the data model. They're details I don't build into the libary's exposed API. They can be used internally and externally but never directly. For example I have something to the effect of this in my library:

```C
// GNUC compilers mandated so this is not a problem
typedef union __attribute__((transparent_union))
{
    int_least64_t fsp; // File system position
   uint_least64_t abs; // Platforms where the registers are alias for the same register will always see the full pointer via this
            void *ptr;
      void const *cptr;
           uchar *top;
     uchar const *ctop;
} abs_t;

API_CB abs_t mrawcpy( abs_t dst, abs_t src, abs_t cpy );

#ifndef _LIB
inline size_t lrawcpy( void *dst, void const *src, size_t cpy )
    { return mrawcpy( dst, src, cpy ).abs; }
# ifdef COMPILE_APP_LOCALS
extern inline size_t lrawcpy( void *dst, void const *src, size_t cpy );
# endif
#else
inline size_t lmrawcpy( void *dst, void const *src, size_t cpy )
    { return mrawcpy( dst, src, cpy ).abs; }
# ifdef COMPILE_LIB_LOCALS
extern inline size_t lmrawcpy( void *dst, void const *src, size_t cpy );
# endif
#endif
```
Unless the inline in question does not use the data model in anyway it will be compiled as private to the software being compiled. So the app cannot access the library's internal inlines and vise versa. The library will always be compiled as native to the platform but the app will start as compiled for the architecture only. *.exe vs *.elf vs etc is combated via the launcher that will come with each version of the library. The launcher expects to find apps built for it's library in 3 places, each with their own purpose.

/.lib/lib<ver>/all_<arch>/<author>/<name>.a
~/.lib/lib<ver>/all_<arch>/<author>/<name>.a
$OTG_DIR/.lib/lib<ver>/all_<arch>/<author>/<name>.a

Alternatively there's:
/.lib/all/<author>/<name>.mak
~/.lib/all/<author>/<name>.mak
$OTG_DIR/.lib/all/<author>/<name>.mak

Or

/.lib/all/<author>/src/<name>/build.mak
~/.lib/all/<author>/src/<name>/build.mak
$OTG_DIR/.lib/all/<author>/src/<name>/build.mak

The library will hand MAIN_O and LIBMAIN_O variables to the makefiles along with the expected ABI path so the final app/lib can be made and passed into the PATH variable.

Basically I'm abusing a lot of common stuff to produce a new software ecosystem that does not need to know the operating system under which their binaries are being run. The option will be there to go native but it will not be needed for most software (such as eog which only needs to be able to open image files and can defer the file browsing prompt to whatever my library linked).

**Comment by awsdert** [comment]
btw I don't really care which of the fixed width integer sizes you use so long as it's fixed width.

**Comment by floooh** [comment]
> long can be 32 or 64 bit,
> int can be 16, 32 or 64 bit
> short can be 16 or 64 bit

Theoretically that's true, but the more important question is: are there any relevant platforms and compilers where `int` isn't a 32-bit signed integer? The only platforms I'm aware of are 16-bit platforms like Intel 80286, some early Amiga compilers and embedded 8-bit platforms like the Arduino, but the sokol headers don't support any of those platforms anyway.

It's the same as little- vs big-endian. In theory code should be endian agnostic, but in practice little-endian has won and big-endian is an exotic edge case that's not worth supporting anymore.

**Comment by awsdert** [comment]
The backend of sokol would still be implementation defined in those cases but having a standard expectation of ABI built into the API would force those implementations to at least present the results in the expected sizes, even if it doesn't need those sizes for what it gives. This library and launcher of mine is aimed at **any** platform, that includes consoles from way back when like the nintendo gameboy. Not expected to be used fully in the final product for sure but it is expecting to at least make the initial development easier by allowing devs to develop on native platform then gradually prune the result until it fits on the target console's memory (since pruning would just mean reducing functions and data needed).

**Comment by voidware** [comment]
This would be a world of hurt. And almost no gain.

**Comment by awsdert** [comment]
Not really, it's mainly just changing from variable sized integers (short, int, long) to fixed size integers. For anything native (such as the backend for instance) you would just slap something like `#ifdef USE_NATIVE_SOKOL` around it. Doesn't have to be all in one shot either, just something to help you start a programming session for instance (since the 1st 15min is the hardest to start). you can add a prefix too for where you want to draw the line between ABI and API. In my wrapper I'm using `mpaw`/`'lmpaw` for module/library functions, inlines & typedefs (which is where the ABI stuff is handled); `paw` for current binary stuff (everything is an inline to the mpaw stuff, data model is handled via a static object that the inlines pass onto the mpaw calls), and I'm using `npaw` for anything fully native (like `npawfso_open()`) which passes unverified parameters straight onto `open()`, `CreateFileA()` or `CreateFileW`. There's plenty to be gained with this approach because it directly removes unexpected behaviour caused by unnoticed assumptions about the ABI. If the ABI is well defined then you can force the middle layer (the defined ABI) to have optimised behaviour.

For example let's say you have something like `void sokol_addivec4( int a[4], int b[4] ) {...}` The optimisations applied on the assumption int is 4 bytes can lead to buggy behaviour on platforms where it's not (such as the LP32/ILP64/SILP64 models) whereas if it's say `void msokol_addivec4( int32_t a[4], int32_t b[4] )` you can inline the former to `msokol_addsvec4(...)` in the LP32 model and `msokol_addlvec4(...)` in the ILP64/SILP64 models and know that even if `sokol_addivec4(...)` behaves wrong on some new data model it's not because of the functions it's inlined to but because the inline is mapping the wrong function on that model.

In short bug fixing becomes easier if you've already ruled out the size of the integers, structures, etc effecting the implementation.

**Comment by voidware** [comment]
As you point out `int` is not fixed size, but neither is `int_least32_t` so I can't see that changing to these can _ever_ be useful _unless_ you're working on systems where `int` is _less than_ 32 bits.

Only retro machines had that, and they wouldn't be using Sokol for graphics.

**Comment by awsdert** [comment]
The variable width of int_least*_t isn't a problem, it just helps me to create wrapper calls when I get round to the graphics side of things. My types are fixed in size via the `_BitInt` type so I only need to concern myself with what's the maximum the source API could be taking on any system. If the maximum is 32 bits then I can just declare pawe32d as the size the wrapper uses. On the other hand if you don't define those max width expectations then I have to use pawrd which is defined as the register width and cannot optimise array sizes as a result.

**Comment by floooh** [comment]
I'll close this ticket as 'not planned'. You can safely assume that int is 32 bits on all platforms supported by sokol_gfx.h.
